helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/constraints/ConstraintBasedAlgorithm.java

```diff
@@ -68,23 +71,47 @@ public OptimalAssignment calculate(ClusterModel clusterModel) throws HelixRebala
     List<AssignableNode> nodes = new ArrayList<>(clusterModel.getAssignableNodes().values());
     Set<String> busyInstances =
         getBusyInstances(clusterModel.getContext().getBestPossibleAssignment().values());
-    // Sort the replicas so the input is stable for the greedy algorithm.
-    // For the other algorithm implementation, this sorting could be unnecessary.
-    for (AssignableReplica replica : getOrderedAssignableReplica(clusterModel)) {
+    List<AssignableReplica> toBeAssignedReplicas = clusterModel.getAssignableReplicaMap().values().stream()
+        .flatMap(replicas -> replicas.stream()).collect(Collectors.toList());
+    // Compute overall utilization of the cluster. Capacity dimension -> <total usage, total capacity>
+    Map<String, MutablePair<Integer, Integer>> overallClusterUtilMap =
+        computeOverallClusterUtil(nodes);
+    Map<String, Integer> replicaHashCodeMap = toBeAssignedReplicas.parallelStream().collect(Collectors
+        .toMap(AssignableReplica::toString,
+            replica -> Objects.hash(replica.toString(), clusterModel.getAssignableNodes().keySet()),
+            (hash1, hash2) -> hash2));
+    Pair<AssignableReplica, List<AssignableReplica>> replicaRes =
+        getNextAssignableReplica(toBeAssignedReplicas, overallClusterUtilMap, clusterModel,
+            replicaHashCodeMap);
+    AssignableReplica replica = replicaRes.getLeft();
+    toBeAssignedReplicas = replicaRes.getRight();
+
+    while (replica != null) {
```

Comment:
Could do-while fit better here?

Reply:
I am not sure... I think we still need to call `getNextAssignableReplica` twice when using do-while .

Reply:
How about 
```
do {
...
}
while (!toBeAssignedReplicas.isEmpty())
```
?

Reply:
Thanks for the offline discussion. Changed to use `toBeAssignedReplicas.isEmpty()` in while evaluate expression.